### PR TITLE
cgroups: add cgroup-name Go helper

### PR DESCRIPTION
**Pull request by ktsaou**
## Summary

- Replace the shell `cgroup-name.sh` helper with a Go `cgroup-name` helper.
- Wire the helper into CMake, installer, and package permission paths.
- Preserve cgroups plugin timeout handling and helper failure behavior with focused tests.

## Split Scope

- This PR contains only the cgroup-name helper replacement surface.
- It is based directly on `master` and does not include topology container attribution changes.

## Validation

- `go test ./...` in `src/collectors/cgroups.plugin/cgroup-name`
- `cmake -S . -B .local/build-cgroup-name-only -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_PLUGIN_GO=ON -DENABLE_PLUGIN_XENSTAT=OFF -DENABLE_PLUGIN_FREEIPMI=Off -DENABLE_PLUGIN_CUPS=Off -DENABLE_PLUGIN_NFACCT=Off -DENABLE_PLUGIN_OTEL=Off -DENABLE_PLUGIN_OTEL_SIGNAL_VIEWER=Off -DENABLE_EXPORTER_PROMETHEUS_REMOTE_WRITE=Off -DENABLE_EXPORTER_MONGODB=Off -DENABLE_BUNDLED_PROTOBUF=Off`
- `cmake --build .local/build-cgroup-name-only --target cgroup-name-target netdata -j "$(nproc)"`

## Upgrade Notes

- **Kubernetes API TLS is now verified by default.** The previous shell helper used `curl -k` and never verified the Kubernetes API-server certificate. The Go helper verifies it against the in-cluster CA (`/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`), matching `client-go` and the go.d Kubernetes collectors. Standard clusters (kubeadm/EKS/GKE/AKS) are unaffected. On clusters where the API endpoint does not chain to that CA (custom PKI, a TLS-intercepting proxy, or an external load balancer with its own certificate), Kubernetes pods/containers fall back to generic `k8s_<id>` names until you set `K8S_TLS_INSECURE=true` in the Netdata service environment. This can change chart identity for those nodes on upgrade.
- **The `[plugin:cgroups] script to get cgroup names` option was removed.** The helper path is now fixed at `/usr/libexec/netdata/plugins.d/cgroup-name`; any custom value is ignored.
- **The default `[plugin:cgroups] cgroup-name timeout` is now 120s** (was 15s; the old shell helper had no timeout). Set `0` to disable.

---
## Summary by cubic
Replaces the shell `cgroup-name.sh` with a Go `cgroup-name` binary and wires it into build, installers, packages, CI, and runtime. Hardens resolution, cache I/O, and TLS; restores fallback behavior; raises the default timeout to 120s; updates docs; and migrates the legacy helper path.

- **Refactors**
  - Added Go `cgroup-name`; removed `cgroup-name.sh`; new `ENABLE_CGROUP_NAME`; installed as `/usr/libexec/netdata/plugins.d/cgroup-name` with root:`netdata` 0750; runtime/package checks and overlay-upgrade cleanup.
  - Discovery adds +2s grace, stops retries after timeouts, and applies the exit-2 fallback name on the last retry; default `cgroup-name timeout` is 120s.
  - Resolves `docker`/`kubectl`/`podman`/`ps` via PATH; dropped the `jq` gate; removed unreachable docker/podman CLI fallbacks; clearer `K8S_TLS_INSECURE` guidance.
  - Secured caches and labels: atomic 0600 temp-file writes + rename; reject non-regular files; single-line container-id cache matches; precompiled dispatch regexes; extracted name/label parsing into `cgroup-name-labels.{c,h}` with tests; trimmed service-account token and GCP metadata values; CI runs the standalone Go module plus `cgroups-plugin-labels-test` and `cgroup-name-config-test`.

- **Migration**
  - The legacy `script to get cgroup names = /usr/libexec/netdata/plugins.d/cgroup-name.sh` is auto-migrated to `/usr/libexec/netdata/plugins.d/cgroup-name`; the config key remains and custom paths continue to work.
  - Configure the helper with `cgroup-name timeout` in `[plugin:cgroups]` (default `120000` ms; set `0` to disable).
  - Kubernetes API TLS is verified by default; set `K8S_TLS_INSECURE=true` if your cluster uses a non-standard CA or an intercepting proxy.

<sup>Written for commit 7d9468fc40cab62dd061b2e3c4342f439f1d706e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/22685?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+	}
+	return true
+}
+
+var k8sTLSInsecureWarned sync.Once
+
+func (r *resolver) k8sTLSConfig(mode k8sTLSMode) *tls.Config {
+	switch mode {
+	case tlsModeKubelet:
+		return &tls.Config{MinVersion: tls.VersionTLS12, InsecureSkipVerify: true}
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in 8e89039982. This thread is outdated: the TLS policy is now explicit in code. API-server calls verify against the service-account CA by default, while kubelet calls keep the legacy-compatible insecure mode because common kubelet serving certs are self-signed or do not verify for https://localhost:10250. The insecure kubelet path is documented in code with the compatibility rationale and NOSONAR annotation.

[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in 8e89039982. This thread is outdated: kubelet requests intentionally keep the legacy-compatible insecure TLS mode because https://localhost:10250 commonly cannot pass hostname verification with stock kubelet serving certificates. API-server requests still verify by default; only the kubelet compatibility path and the explicit K8S_TLS_INSECURE escape hatch disable verification, both with source comments and NOSONAR rationale.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 2]
```diff
@@ -0,0 +1,1353 @@
...
+	case tlsModeAPIServer:
+		if k8sTLSInsecure() {
+			k8sTLSInsecureWarned.Do(func() {
+				r.warning("K8S_TLS_INSECURE is set: TLS verification of Kubernetes API calls is disabled")
+			})
+			return &tls.Config{MinVersion: tls.VersionTLS12, InsecureSkipVerify: true}
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in 8e89039982. This thread is outdated: API-server TLS now verifies by default through the mounted Kubernetes service-account CA. The only API-server path that disables certificate verification is the explicit operator escape hatch K8S_TLS_INSECURE, and that branch now carries a warning plus source-level rationale/NOSONAR annotation.

[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in 8e89039982. This thread is outdated: API-server hostname verification remains enabled by default via the service-account CA TLS config. Hostname verification is skipped only when an operator explicitly sets K8S_TLS_INSECURE for clusters whose API endpoint cannot verify against the mounted CA; that compatibility escape hatch is now documented in code and logs a warning.

File: src/collectors/cgroups.plugin/cgroup-discovery.c [hunk 3]
```diff
@@ -245,25 +300,65 @@ static inline void discovery_rename_cgroup(struct cgroup *cg) {
...
+    int wait_ms = -1;
+    if (cgroup_name_timeout_ms > 0)
+        wait_ms = cgroup_name_timeout_ms > INT_MAX - CGROUP_NAME_GRACE_MS ?
+                      INT_MAX :
+                      cgroup_name_timeout_ms + CGROUP_NAME_GRACE_MS;
+
+    struct pollfd pfd = { .fd = spawn_popen_read_fd(instance), .events = POLLIN };
+    int pr;
+    do {
+        pr = poll(&pfd, 1, wait_ms);
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-discovery.c]: Fixed in 4503c9aff0: the parent now computes one monotonic deadline for the helper wait and, after EINTR, polls only the remaining time. timeout=0 still keeps the legacy unbounded wait behavior.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 4]
```diff
@@ -0,0 +1,1456 @@
...
+	}
+	return true
+}
+
+var k8sTLSInsecureWarned sync.Once
+
+func (r *resolver) k8sTLSConfig(mode k8sTLSMode) *tls.Config {
+	switch mode {
+	case tlsModeKubelet:
+		return &tls.Config{
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in d9c5d39c5d. This is the kubelet compatibility path: kubelet serving certificates are commonly self-signed or not valid for https://localhost:10250, so this preserves the approved legacy-compatible behavior. The NOSONAR rationale is now on the exact Sonar-reported struct literal and field lines.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 5]
```diff
@@ -0,0 +1,1456 @@
...
+			MinVersion:         tls.VersionTLS12,
+			InsecureSkipVerify: true, // NOSONAR - kubelet serving certs are commonly self-signed; this preserves legacy helper compatibility.
+		}
+	case tlsModeAPIServer:
+		if k8sTLSInsecure() {
+			k8sTLSInsecureWarned.Do(func() {
+				r.warning("K8S_TLS_INSECURE is set: TLS verification of Kubernetes API calls is disabled")
+			})
+			return &tls.Config{
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in d9c5d39c5d. API-server TLS still verifies by default through the service-account CA; this branch only applies when an operator explicitly sets K8S_TLS_INSECURE for clusters that cannot verify against that CA. The escape hatch logs a warning, and the NOSONAR rationale is now on the exact Sonar-reported lines.

[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Addressed in d9c5d39c5d. Hostname verification remains enabled for API-server calls by default. It is skipped only for the explicit K8S_TLS_INSECURE operator escape hatch, which is documented in code, logs a warning, and now carries NOSONAR rationale on the exact reported lines.

File: src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+run_one "$go_resolver" "$tmpdir/go.out" "$tmpdir/go.err" "$tmpdir/go.code" "$tmpdir/go.ms" "$@"
+
+if cmp -s "$tmpdir/shell.out" "$tmpdir/go.out" &&
+   cmp -s "$tmpdir/shell.code" "$tmpdir/go.code" &&
+   diff -q <(normalize_for_diff "$tmpdir/shell.err") <(normalize_for_diff "$tmpdir/go.err") >/dev/null; then
+  diff_status="match"
+  diff_detail=""
+else
+  diff_status="mismatch"
+  diff_detail="$(diff -u "$tmpdir/shell.out" "$tmpdir/go.out" || true)"
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger]: Fixed in 581ca23cf5. The wrapper now normalizes stderr once, compares stdout/stderr/exit-code independently, and records per-channel diff_detail sections only for the channels that actually mismatch. I smoke-tested the stderr-only plus exit-code mismatch case.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    "$(cat "$tmpdir/shell.ms")"
+  printf '"go":{"stdout":"%s","stderr":"%s","exit_code":%s,"duration_ms":%s},' \
+    "$(json_escape "$(cat "$tmpdir/go.out")")" \
+    "$(json_escape "$(cat "$tmpdir/go.err")")" \
+    "$(cat "$tmpdir/go.code")" \
+    "$(cat "$tmpdir/go.ms")"
+  printf '"diff_status":"%s",' "$diff_status"
+  printf '"diff_detail":"%s"' "$(json_escape "$diff_detail")"
+  printf '}\n'
+} >>"$capture_file"
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger]: Valid — fixed in ca513c16f6. The capture-file write is now best-effort: the JSON block redirects with `>>"$capture_file" 2>/dev/null || true`, so a failed capture write can no longer abort the wrapper before it returns the resolver result. (This wrapper is developer-only parity tooling and is not installed, but it must stay a transparent stand-in for `cgroup-name.sh.real`.)

File: src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+cat "$tmpdir/shell.out"
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/capture-wrappers/cgroup-name.sh.logger]: Valid — fixed in ca513c16f6. The wrapper now relays the shell resolver stderr verbatim (`cat "$tmpdir/shell.err" >&2`) before passing through stdout and exiting with the resolver exit code, so resolver diagnostics/logs are no longer swallowed.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 9]
```diff
@@ -0,0 +1,1524 @@
...
+	return name, 1
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+func writePrivateFile(path string, data []byte) error {
+	if err := os.WriteFile(path, data, 0o600); err != nil {
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Valid. The Kubernetes cache files use fixed, predictable names in a world-writable TMPDIR (`/tmp/netdata-cgroups-*`), so a planted symlink could redirect a root-owned write. Fixed in ca513c16f6: `writePrivateFile` now writes atomically via `os.CreateTemp` (O_EXCL, mode 0600) + `os.Rename`, which replaces a destination symlink instead of following it. I also closed the matching read vector — `firstLineFile`/`grepFile` and the cache-presence gate now reject non-regular files via `isPrivateRegularFile` (Lstat). Covered by new `TestWritePrivateFileIsSymlinkSafe` and `TestCacheReadHelpersRejectSymlinks`.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 10]
```diff
@@ -0,0 +1,1524 @@
...
+			id = m[2]
+		}
+		r.podmanValidateID(ctx, id, cgroup)
+	case reNspawn.MatchString(cgroup):
+		r.name = reNspawn.ReplaceAllString(cgroup, "$1")
+	case strings.Contains(cgroup, "machine.slice_machine") && strings.Contains(cgroup, "-lxc"):
+		r.name = "lxc/" + lxcMachineName(cgroup)
+	case strings.Contains(cgroup, "machine.slice_machine") && strings.Contains(cgroup, "-qemu"):
+		r.name = "qemu_" + machineName(cgroup, "-qemu")
+	case regexp.MustCompile(`machine_.*\.libvirt-qemu`).MatchString(cgroup):
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Valid — fixed in ca513c16f6. Precompiled the pattern at package scope as `reLibvirtQemu`, matching the other matcher regexes. Also fixed the same class nearby (`reLxcPayload`, `reProxmoxConfName`, `reProxmoxConfHost`) and reused the existing `reMachineIDSegment` in `lxcMachineName`, which was recompiling the identical pattern inline.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again — pushed ca513c16f6 addressing the 4 open findings (best-effort capture + stderr relay in the dev wrapper, atomic symlink-safe cache writes plus symlink-rejecting reads with tests, and package-scope precompiled dispatch regexes), and rebased onto current master.

**Comment by ktsaou** [comment]
@ilyam8 this has been reviewed extensively. With your review we should merge it.
The key difference is that now k8s API is verified.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: packaging/makeself/install-or-update.sh [hunk 11]
```diff
@@ -187,7 +187,11 @@ fi
 
 progress "changing plugins ownership and permissions"
 
+<<<<<<< HEAD
 for x in ndsudo apps.plugin perf.plugin slabinfo.plugin debugfs.plugin freeipmi.plugin ioping cgroup-network local-listeners network-viewer.plugin ebpf.plugin ebpf-go.plugin nfacct.plugin xenstat.plugin python.d.plugin charts.d.plugin go.d.plugin snmp-trap-profile-gen ioping.plugin cgroup-network-helper.sh otel-plugin otel-signal-viewer-plugin systemd-journal.plugin macos-logs.plugin netflow-plugin; do
+=======
+for x in ndsudo apps.plugin perf.plugin slabinfo.plugin debugfs.plugin freeipmi.plugin ioping cgroup-network local-listeners network-viewer.plugin ebpf.plugin ebpf-go.plugin nfacct.plugin xenstat.plugin python.d.plugin charts.d.plugin go.d.plugin snmp-trap-profile-gen ioping.plugin cgroup-network-helper.sh cgroup-name otel-plugin otel-signal-viewer-plugin systemd-journal.plugin netflow-plugin; do
+>>>>>>> 90419a71ac (cgroups: add cgroup-name Go helper)
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply packaging/makeself/install-or-update.sh]: Valid catch - the conflict markers from the rebase were committed by mistake. Fixed in 682b5fa88d: the plugin ownership list is now a single line containing both master's macos-logs.plugin and the new cgroup-name helper, and the fix is applied in the commit that introduced the conflict, so no commit in the stack carries markers. Swept the whole tree for leftover markers (rg '^<<<<<<<|^>>>>>>>') across all 12 commits - this was the only instance. bash -n passes.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 12]
```diff
@@ -0,0 +1,1504 @@
...
+		base = "https://localhost:10250"
+	}
+	return base + "/pods"
+}
+
+func (r *resolver) k8sFetchPods(ctx context.Context, fn, kubeSystemUID string) (string, string, int) {
+	if os.Getenv("KUBERNETES_SERVICE_HOST") != "" && os.Getenv("KUBERNETES_PORT_443_TCP_PORT") != "" {
+		tokenBytes, _ := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/token")
+		header := map[string]string{"Authorization": "Bearer " + string(tokenBytes)}
+		host := os.Getenv("KUBERNETES_SERVICE_HOST") + ":" + os.Getenv("KUBERNETES_PORT_443_TCP_PORT")
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Valid - fixed in abd5a1ccb0. The token is now read via a package-level path var and trimmed with strings.TrimSpace before building the Authorization header (the shell's $(<file) stripped trailing newlines; net/http rejects header values containing a newline outright, so this was a hard failure, not just an auth mismatch). Swept the diff for the same class: the GCP metadata values had the identical issue (shell $(curl ...) also newline-stripped and tested the stripped value with -n) and are now trimmed too, with a whitespace-only body treated as a failed fetch. The API-server test now seeds a token file WITH a trailing newline and asserts the exact Authorization header the server receives.

File: src/collectors/cgroups.plugin/cgroup-discovery.c [hunk 13]
```diff
@@ -250,25 +211,91 @@ static inline void discovery_rename_cgroup(struct cgroup *cg) {
...
+    int exit_code = -1;
+    if (pr <= 0) {
+        // timeout or poll error: the helper hung. Kill it, drop any partial
+        // output, and stop retrying this cgroup - re-running a helper that
+        // already hung would just block discovery again every cycle. The cgroup
+        // keeps its raw chart id.
+        collector_error(
+            "CGROUP: rename helper for '%s' did not respond within %d ms; killing it.", cg->id, wait_ms);
+        spawn_popen_kill(instance, 0);
+        cg->pending_renames = 0;
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-discovery.c:246]: This is a deliberate, recorded design decision, not an accident. The old helper had NO timeout at all: one hung curl/kubectl call blocked discovery forever. The retry ladder (pending_renames 9/2) is fully preserved for the normal case - exit-2 responses keep retrying exactly as before. The suppression applies ONLY when the helper exceeded the operator timeout (default 120s) plus grace and had to be killed: re-running a helper that just proved it can hang would block discovery for another 120s every cycle, indefinitely, on every affected cgroup. With the generous 120s budget a genuine transient that long is rare, and the cost asymmetry (one cgroup keeps its raw id vs discovery starving for all cgroups) favors stopping. The comment above the branch documents this trade-off.

File: src/collectors/cgroups.plugin/cgroup-name/FLOW.md [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+The Kubernetes cache files are:
+
+- `${TMPDIR:-/tmp}/netdata-cgroups-k8s-cluster-name`
+- `${TMPDIR:-/tmp}/netdata-cgroups-kubesystem-uid`
+- `${TMPDIR:-/tmp}/netdata-cgroups-containers`
+
+For container ids, all three files must exist and the containers file must have
+a grep match for the container id to use the cache. Otherwise the binary reads
+any existing cluster/system uid cache, discovers missing values, then overwrites
+the cache files with non-atomic writes equivalent to `echo > file 2>/dev/null`.
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/FLOW.md]: Valid - fixed in f56d0697dc. FLOW.md now describes the shipped behavior: atomic writes via an unguessable 0600 temp file in the same directory renamed over the destination (symlink at the predictable path is replaced, not followed), and notes the shell's plain echo > file as the legacy behavior. The same commit also fixed two other stale shell-era claims in FLOW.md (the jq PATH gate, which the Go helper does not have, and the missing note that inspect output is parsed rather than eval'ed).

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 15]
```diff
@@ -0,0 +1,1504 @@
...
+	}
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return false
+	}
+	procs := strings.Fields(string(data))
+	if len(procs) != 1 {
+		return false
+	}
+	comm, err := os.ReadFile(filepath.Join("/proc", procs[0], "comm"))
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Intentional bug-for-bug parity with the shell helper, which reads /proc/${procs[0]}/comm unprefixed at cgroup-name.sh.in line 257 - this PR's contract is behavioral equivalence with that helper. In supported containerized deployments (docker run --pid=host, the k8s helm chart) the plugin runs in the host PID namespace, so the local /proc resolves host PIDs and the unprefixed read is correct there; prefixing would instead break hosts where NETDATA_HOST_PREFIX mounts /host/sys but not /host/proc identically. Added a source comment in 69d24f7f40 documenting this choice. Whether the pause check should honor NETDATA_HOST_PREFIX is a behavior change affecting the shell helper's logic equally, tracked as a possible follow-up outside this PR.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 16]
```diff
@@ -0,0 +1,1504 @@
...
+		if err != nil {
+			r.warning(fmt.Sprintf("%s: error on curl '%s': %s.%s", fn, url, err.Error(), tlsHint(err)))
+			return kubeSystemNS, "", 1
+		}
+		return kubeSystemNS, string(body), 0
+	}
+
+	if r.kubeletProcessRunning(ctx) {
+		if commandAvailable("kubectl") {
+			kubeConfig := os.Getenv("KUBE_CONFIG")
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Intentional bug-for-bug parity with the shell helper: it ran the kube-system namespace kubectl call with --kubeconfig="$KUBE_CONFIG" (empty when unset, cgroup-name.sh.in line 427) and only applied the admin.conf default at line 432, between the two calls, so the pods query got the fallback and the namespace query did not. The Go code replicates that ordering deliberately - changing it would alter which kubeconfig the namespace query uses on hosts that rely on kubectl's default loading rules. Added a source comment in 69d24f7f40 explaining the quirk so it reads as intent, not oversight.

File: packaging/cmake/Modules/NetdataGoTools.cmake [hunk 17]
```diff
@@ -33,15 +34,21 @@ set(GO_LDFLAGS "${GO_LDFLAGS} -X ${BUILDINFO_PKG}.LogDir=${LOG_DIR}")
 # and then appending the go.mod and go.sum files from the root of the
 # source tree.
 macro(add_go_target target output build_src build_dir)
+    cmake_parse_arguments(ARG "" "EXTRA_LDFLAGS" "" ${ARGN})
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply packaging/cmake/Modules/NetdataGoTools.cmake]: Addressed in 69d24f7f40. Cross-call contamination cannot happen (cmake_parse_arguments undefines not-given keywords at each invocation), but the scratch variables did linger in the including scope after the last expansion, so the macro now unsets ARG_EXTRA_LDFLAGS, ARG_UNPARSED_ARGUMENTS, ARG_KEYWORDS_MISSING_VALUES, and ${target}_GO_LDFLAGS on exit. Kept macro() rather than converting to function(): the ${target}_DEPS globbing pattern predates this PR and every existing caller uses the macro form; converting the shared build macro is beyond this PR's scope. CMake reconfigure and the cgroup-name target build verified after the change.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 18]
```diff
@@ -0,0 +1,1520 @@
...
+	defer r.track("snap-list-docker", time.Now())
+	return exec.CommandContext(ctx, "snap", "list", "docker").Run() == nil
+}
+
+func (r *resolver) dockerGetName(ctx context.Context, id string) {
+	if r.snapHasDocker(ctx) {
+		r.dockerLikeGetNameAPI(ctx, "DOCKER_HOST", id)
+	} else if commandAvailable("docker") {
+		r.dockerLikeGetNameCommand(ctx, "docker", id)
+	} else if !r.dockerLikeGetNameAPI(ctx, "DOCKER_HOST", id) {
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Correct observation, wrong conclusion about a regression: the shell had the same effective behavior. In the shell, DOCKER_HOST/PODMAN_HOST are always defaulted with :=, and a failed curl still ended in 'return 0', so the '|| podman inspect' CLI leg could only ever run when jq was missing. With jq present (the documented requirement for API resolution) the shell never fell back to the podman CLI on an API failure - exactly like this code. Since the Go helper parses JSON natively, the jq-missing trigger no longer exists, making the CLI leg dead code rather than a lost feature. Fixed in 59c8e01b68 by removing the unreachable legs outright (docker branch 3 and the podman flow are now explicitly API-only, matching the shell's jq-present behavior), dropping the now-meaningless boolean return, and updating FLOW.md. Adding a real API-failure fallback to the podman CLI would be a new behavior the shell never had - out of scope for this parity port.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 19]
```diff
@@ -0,0 +1,1524 @@
...
+	return err == nil && !st.IsDir()
+}
+
+func firstConfigValue(path string, sedRE *regexp.Regexp, grepPrefix string) string {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return ""
+	}
+	for line := range strings.SplitSeq(string(data), "\n") {
+		if !strings.HasPrefix(line, strings.TrimPrefix(grepPrefix, "^")) {
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: False positive for this parity port: the mismatch is inherited from the shell verbatim. The shell pipeline is grep -e '^name: ' FILE | head -1 | sed -rn 's|\\s*name\\s*:\\s*(.*)?$|\\1|p' (cgroup-name.sh.in lines 711/719) - the strict '^name: ' grep gate runs BEFORE the whitespace-lenient sed, so a name:vm line (no space) is skipped by the shell exactly as it is here. firstConfigValue encodes that pipeline literally (grepPrefix = the grep, sedRE = the sed); Proxmox itself always writes 'name: value'. Added a comment on the function in d2080ab9ae so the strictness reads as intent.

File: src/collectors/cgroups.plugin/cgroup-name/main.go [hunk 20]
```diff
@@ -0,0 +1,1524 @@
...
+		label := labelPair(lname, inspectLabelValue(lval))
+		if r.labels != "" {
+			r.labels += "," + label
+		} else {
+			r.labels = label
+		}
+	}
+}
+
+func (r *resolver) dockerLikeGetNameCommand(ctx context.Context, command, id string) bool {
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply src/collectors/cgroups.plugin/cgroup-name/main.go]: Valid - fixed in d2080ab9ae. The always-true return lost its last consumer when the unreachable podman CLI legs were removed; the function now returns nothing, matching its API sibling. Swept the file for other unconsumed boolean returns - every remaining bool-returning helper is branched on at its call sites.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

**Comment by ilyam8** [comment]
@copilot resolve the merge conflicts in this pull request

**Review by ilyam8 — APPROVED** [review]
